### ISSUE TITLE
 Support CurveBS metadata stored in the database

### ISSUE THREAD

**Issue by ilixiaocui** [opencurve/curve#2224]
### Description

Currently the [metadata](https://github.com/opencurve/curve/blob/master/src/mds/nameserver2/namespace_storage.cpp) of current CurveBS is stored in [Etcd](https://github.com/opencurve/curve/tree/master/src/kvstorageclient). In order to apply to more scenarios, it is hoped that the metadata can add a `sql database` like mysql, MariaDB, PostgreSQL, etc. as one of the storage engines.

It is hoped that the user can choose whether the metadata is stored in the kv engine or the sql engine through the configuration file.

### Proposed solution

Add a new `sqlstorageclient` for [metadata](https://github.com/opencurve/curve/blob/master/src/mds/nameserver2/namespace_storage.cpp) . 
It should be noted that in addition to storing metadata, Etcd also undertakes the function of [leader election](https://github.com/opencurve/curve/tree/master/src/leader_election), and this part also needs to be modified.

**Comment by Writer-X** [comment]
Hello Mr./Ms. Li, I'm new to open source and saw this project in GSoC 2023.  Could you tell me something about the difficulties of this project and the specific requirements for contributors?

**Comment by Cyber-SiKu** [comment]
@ilixiaocui

**Comment by ilixiaocui** [comment]
> Hello Mr./Ms. Li, I'm new to open source and saw this project in GSoC 2023. Could you tell me something about the difficulties of this project and the specific requirements for contributors?

Thanks for following this issue. The issue has already described the places that need to be modified. Is the description in the issue clear to you?

**Comment by Writer-X** [comment]
> > Hello Mr./Ms. Li, I'm new to open source and saw this project in GSoC 2023. Could you tell me something about the difficulties of this project and the specific requirements for contributors?
> 
> Thanks for following this issue. The issue has already described the places that need to be modified. Is the description in the issue clear to you?

Thanks for your reply.   
In my understanding, taking mysql as an example, the issue needs to implement some interfaces to mysql and implement a leader election mechanism for multiple mysql nodes.  
My major is closely related to data processing, and I am interested in this issue.  If I want to contribute to it, should I prepare something other than reading the source code and trying to run the program?

**Comment by yyy1000** [comment]
I'm a senior student in China and am interested in this. Please kindly let me know if any doc helps me to get started.

**Comment by ilixiaocui** [comment]
> > > Hello Mr./Ms. Li, I'm new to open source and saw this project in GSoC 2023. Could you tell me something about the difficulties of this project and the specific requirements for contributors?
> > 
> > 
> > Thanks for following this issue. The issue has already described the places that need to be modified. Is the description in the issue clear to you?
> 
> Thanks for your reply. In my understanding, taking mysql as an example, the issue needs to implement some interfaces to mysql and implement a leader election mechanism for multiple mysql nodes. My major is closely related to data processing, and I am interested in this issue. If I want to contribute to it, should I prepare something other than reading the source code and trying to run the program?

There are no others for now.  Would you like to join the contributor club for further communication？Please note the issue id when joining the group if you want.
<img width="380" alt="image" src="https://user-images.githubusercontent.com/15689619/223030108-c75fbce8-798b-49f6-a01a-9e237ca06289.png">

**Comment by ilixiaocui** [comment]
> I'm a senior student in China and am interested in this. Please kindly let me know if any doc helps me to get started.

How much do you know about the requirements of the current topic? Want to know what kind of documentation you need?

**Comment by Tangruilin** [comment]
1. `Etcd has limited scalability and the amount of metadata that can be stored is limited`, If the problem is `etcd`, why we chose to support a `sql database` but not a better `kv database`.
2. As the description of this issue, I realize that it is to add a new client for `sql database` and give users choices, is it right?

**Comment by ilixiaocui** [comment]
> 1. `Etcd has limited scalability and the amount of metadata that can be stored is limited`, If the problem is `etcd`, why we chose to support a `sql database` but not a better `kv database`.
> 2. As the description of this issue, I realize that it is to add a new client for `sql database` and give users choices, is it right?

1.  a better` kv database` is ok. `sql database` is a functional requirement
2. yes. Thanks for your review, I will change the description to be more accurate.

**Comment by GauriBhandari** [comment]
Hey @ilixiaocui  I am a student from India and I would like to contribute to this issue. I wanted to know if this is still available? If it is available then I would like to write a proposal for this.

**Comment by ilixiaocui** [comment]
> Hey @ilixiaocui I am a student from India and I would like to contribute to this issue. I wanted to know if this is still available? If it is available then I would like to write a proposal for this.

Of course, please note the [timeline](https://developers.google.com/open-source/gsoc/timeline), and feel free to raise any ideas or questions you may have.

**Comment by GauriBhandari** [comment]
> > Hey @ilixiaocui I am a student from India and I would like to contribute to this issue. I wanted to know if this is still available? If it is available then I would like to write a proposal for this.
> 
> Of course, please note the [timeline](https://developers.google.com/open-source/gsoc/timeline), and feel free to raise any ideas or questions you may have.

Hey @ilixiaocui  I am unable to see this organization (project) on gsoc organization dashboard. Is it still available? Could you please help

**Comment by bit-dance** [comment]
hello@GauriBhandari .This organization is under cncf and you can search cncf and find their idea lists.Hope it is helpful to you.

**Comment by aavhad1910** [comment]
Hi @ilixiaocui and @bit-dance, I am willing to submit a proposal as solutions for this issue, but I see that this issue has already been assigned to someone. Is this assignment final? If not, can I directly submit a proposal?

**Comment by ilixiaocui** [comment]
> Hi @ilixiaocui and @bit-dance, I am willing to submit a proposal as solutions for this issue, but I see that this issue has already been assigned to someone. Is this assignment final? If not, can I directly submit a proposal?

@bit-dance Sorry, the registration period has ended.

**Comment by ilixiaocui** [comment]
this topic is in Curve's [GSoC2023](https://github.com/cncf/mentoring/discussions/918) and has passed the registration phase. 

welcome to pay attention to other [good first issues](https://github.com/opencurve/curve/issues?q=is%3Aopen+label%3A%22good+first+issue%22+-label%3Aassigned) .

**Comment by zztaki** [comment]
Mark, I want to learn a lot from this modify~
